docs/storage/commit_log/signer/signer.go

```diff
@@ -0,0 +1,174 @@
+		if ourSTH == nil {
+			glog.Errorf("%s: local DB has data ahead of STHs topic!!", s.Name)
+			return
+		}
+		if ourSTH.Offset != dbSTHInfo.sthOffset {
+			glog.Errorf("%s: local DB recorded offset %d but that has inconsistent STH %s!!", s.Name, dbSTHInfo.sthOffset, ourSTH)
+			return
+		}
+		if ourSTH.TimeStamp != dbSTHInfo.sth.TimeStamp || ourSTH.TreeSize != dbSTHInfo.sth.TreeSize {
+			glog.Errorf("%s: local DB has different data than STHs topic!!", s.Name)
+			return
+		}
+		glog.V(2).Infof("%s: our DB at %v, matches STH at that offset", s.Name, dbSTHInfo.sthOffset)
+	}
+
+	// Look to see if anyone else has already stored data just ahead of our STH.  This will
+	// normally be the next entry, but we need to ignore any entries that have inconsistent
+	// offsets.
+	nextOffset := dbSTHInfo.sthOffset
+	var nextSTH *STH
+	for {
+		nextOffset++
+		nextSTH = sthFromString(simkafka.Read("STHs/<treeID>", nextOffset))
+		if nextSTH != nil && nextSTH.Offset != nextOffset {
+			// Ignore this one
+			glog.V(2).Infof("%s: ignoring inconsistent STH %s at offset %d", s.Name, nextSTH.String(), nextOffset)
+			continue
+		}
+		break
+	}
+
+	time.Sleep(*pessimizeInterval)
+	if nextSTH == nil {
+		// We're up-to-date with the STHs topic (as of a moment ago) ...
+		if !s.IsMaster() {
+			glog.V(2).Infof("%s: up-to-date with STHs but not master, so exit", s.Name)
+			return
+		}
+		// ... and we're the master. Move the STHs topic along to encompass any unincorporated leaves.
+		offset := dbSTHInfo.sth.TreeSize
+		batch := simkafka.ReadMultiple("Leaves/<treeID>", offset, *batchSize)
+		glog.V(2).Infof("%s: nothing at next offset %d and we are master, so read %d more leaves", s.Name, nextOffset, len(batch))
+		if len(batch) == 0 {
+			glog.V(2).Infof("%s: nothing to do", s.Name)
+			return
+		}
+		timestamp := (time.Now().UnixNano() / int64(time.Millisecond)) - s.epoch
+		newSTHInfo := STHInfo{
+			sth: STH{
+				TreeSize:  s.db.Size() + len(batch),
+				TimeStamp: timestamp,
+				Offset:    nextOffset, // The offset we expect this to end at
+			},
+			treeRevision: dbSTHInfo.treeRevision + 1,
+		}
+		newSTHInfo.sthOffset = simkafka.Append("STHs/<treeID>", newSTHInfo.sth.String())
+		if newSTHInfo.sthOffset != nextOffset {
+			// The STH didn't get stored at the offset we expected, presumable because someone else got there first
+			glog.Warningf("%s: stored new STH %s at offset %d, which is unexpected; give up", s.Name, newSTHInfo.sth.String(), newSTHInfo.sthOffset)
+			return
+		}
+		glog.V(2).Infof("%s: stored new STH %s at expected offset, including %d new leaves", s.Name, newSTHInfo.sth.String(), len(batch))
+
+		// Now the STH topic is updated (correctly), do our local DB
+		s.db.AddLeaves(timestamp, nextOffset, batch)
+		s.StoreSTHInfo(newSTHInfo)
+	} else {
+		// There is an STH one ahead of us that we're not caught up with yet.
+		if nextSTH.TimeStamp < dbSTHInfo.sth.TimeStamp || nextSTH.TreeSize < dbSTHInfo.sth.TreeSize {
```

Comment:
I updated the "pseudo"code in the doc btw, this check moved up into the `for{}` while scanning for the next STH.

We can tie the doc+code back together in a follow up PR if you prefer though (maybe pop a TODO in here somewhere if you go that route)